### PR TITLE
npm-ify Action Cable

### PR DESCRIPTION
**Pull request by maclover7**
cc @jeremy

### CODE REVIEW

File: tasks/release.rb [hunk 1]
```diff
@@ -44,13 +44,21 @@
       raise "Could not insert PRE in #{file}" unless $1
 
       File.open(file, 'w') { |f| f.write ruby }
+
+      if framework == "actioncable"
+        Dir.chdir("#{root}/actioncable") do
+          version = version.gsub(/\./).with_index { |s, i| i === 2 ? '-' : s }
```
Comment (maclover7) [inline tasks/release.rb]: `5.0.0.rc1` for example, is not a valid npm package version, so it needs to be changed to `5.0.0-rc1`

File: actioncable/package.json [hunk 2]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "actioncable",
+  "version": "5.0.0-beta4",
```
Comment (vipulnsward) [inline actioncable/package.json]: 😎

File: actioncable/jstest/unit/action_cable.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
+require('../test_helper');
+
+describe('ActionCable', function() {
```
Comment (kaspth) [inline actioncable/jstest/unit/action_cable.js]: I have a feeling we might want this to be CoffeeScript 😁

**Comment by jeremy** [comment]
JS test stubs belong in a separate PR. Let's stick to NPM publishing.

File: actioncable/package.json (hunk 2 above)
Comment (vipulnsward) [inline actioncable/package.json]: This needs to be rc1

File: actioncable/jstest/unit/action_cable.js (hunk 3 above)
Comment (vipulnsward) [inline actioncable/jstest/unit/action_cable.js]: Agree, to reflect CS counterparts. https://github.com/basecamp/trix/tree/master/test/src are good examples for these.

**Comment by maclover7** [comment]
👍 will separate out these changes

File: actioncable/package.json (hunk 2 above)
Comment (maclover7) [inline actioncable/package.json]: I figured the version push would be done by someone on core separate from the PR, but I can push up rc1 if people are okay with that.

**Comment by maclover7** [comment]
Updated PR.

File: actioncable/package.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  "description": "WebSocket framework for Rails.",
```
Comment (vipulnsward) [inline actioncable/package.json]: `Ruby on Rails`

Comment (maclover7) [inline actioncable/package.json]: This is the description in `actioncable.gemspec` -- should I update that too? Wanted to try and keep things consistent to avoid confusion :)

Comment (vipulnsward) [inline actioncable/package.json]: We should have this just here, since its not inherent "Rails" means "The Rails" on JS land.

Comment (maclover7) [inline actioncable/package.json]: 👍 makes sense to me

File: actioncable/package.json [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  ],
+  "author": "David Heinemeier Hansson <david@loudthinking.com>",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/rails/rails/issues"
+  },
+  "homepage": "http://rubyonrails.org/",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  }
```
Comment (jeremy) [inline actioncable/package.json]: Why specify this test script?

File: actioncable/package.json [hunk 6]
```diff
@@ -0,0 +1,28 @@
+{
+  "name": "actioncable",
+  "version": "5.0.0-beta4",
+  "description": "WebSocket framework for Ruby on Rails.",
+  "main": "dist/actioncable.js",
+  "files": [
+    "dist/*.css",
```
Comment (jeremy) [inline actioncable/package.json]: What CSS?

File: tasks/release.rb [hunk 7]
```diff
@@ -44,13 +44,21 @@
       raise "Could not insert PRE in #{file}" unless $1
 
       File.open(file, 'w') { |f| f.write ruby }
+
+      if framework == "actioncable"
```
Comment (jeremy) [inline tasks/release.rb]: Can check for `if File.exist?("#{framework}/package.json")` instead of special-casing AC.

File: tasks/release.rb (hunk 1 above)
Comment (jeremy) [inline tasks/release.rb]: Extract an `#npmify_version` or similar that reliably transforms a RubyGems version to an NPM version. Checking for the 3rd `.` _seems_ brittle, but will prob work fine for our versions.

File: tasks/release.rb [hunk 8]
```diff
@@ -44,13 +44,21 @@
...
+        end
+      end
     end
 
     task gem => %w(update_versions pkg) do
       cmd = ""
       cmd << "cd #{framework} && " unless framework == "rails"
       cmd << "bundle exec rake package && " unless framework == "rails"
       cmd << "gem build #{gemspec} && mv #{framework}-#{version}.gem #{root}/pkg/"
+      cmd << "&& npm publish" if framework == "rails"
```
Comment (jeremy) [inline tasks/release.rb]: Publish to NPM _if_ it's the rails gem? Should be publish to NPM if there's an NPM package?

File: actioncable/.gitignore [hunk 9]
```diff
@@ -1,2 +1,3 @@
 /lib/assets/compiled
 /tmp
+/dist
```
Comment (jeremy) [inline actioncable/.gitignore]: Move to the top—alphabetical order.

File: actioncable/package.json (hunk 2 above)
Comment (jeremy) [inline actioncable/package.json]: Use the current version for the initial PR.

File: tasks/release.rb (hunk 1 above)
Comment (matthewd) [inline tasks/release.rb]: > Checking for the 3rd . seems brittle, but will prob work fine for our versions.

Nope; we want to be able to do 4-part versions for security releases. Need to check for non-digits, if that's what's troubling npm.

Comment (maclover7) [inline tasks/release.rb]: Hmm, let me modify the version transformation.

Comment (maclover7) [inline tasks/release.rb]: With new version transformer:

```
Jonathans-MacBook-Pro:rails jon$ irb
irb(main):001:0> version = "5.0.0.beta1.1"
=> "5.0.0.beta1.1"
irb(main):002:0> puts version.gsub(/\./).with_index { |s, i| i >= 2 ? '-' : s }
5.0.0-beta1-1
=> nil
irb(main):003:0> version = "5.0.0.rc1"
=> "5.0.0.rc1"
irb(main):004:0> puts version.gsub(/\./).with_index { |s, i| i >= 2 ? '-' : s }
5.0.0-rc1
=> nil
irb(main):005:0> 
```

File: actioncable/package.json (hunk 2 above)
Comment (maclover7) [inline actioncable/package.json]: 👍 did a push of v5.0.0-rc1 to npm

**Comment by simi** [comment]
I think that second commit needs better description since it is not global change to v5.0.0-rc1.

**Comment by maclover7** [comment]
Ahh whoops, I think that may have been an npm-triggered commit. Let me squash down. :)

**Comment by schneidmaster** [comment]
@maclover7 Pulling your branch to do some work on this this afternoon :)

File: actioncable/package.json [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+  "version": "5.0.0-rc1",
+  "description": "WebSocket framework for Ruby on Rails.",
+  "main": "dist/actioncable.js",
+  "files": [
+    "dist/*.js"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/rails/rails.git"
+  },
```
Comment (jeremy) [inline actioncable/package.json]: Believe this can be shortened to `"repository": "rails/rails"` (same shortcut syntax as `npm install`)

File: actioncable/package.json [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    "actioncable",
+    "rails"
+  ],
+  "author": "David Heinemeier Hansson <david@loudthinking.com>",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/rails/rails/issues"
+  },
+  "homepage": "http://rubyonrails.org/",
+  "scripts": {}
```
Comment (jeremy) [inline actioncable/package.json]: Omit?

File: actioncable/package.json [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+}
```
Comment (jeremy) [inline actioncable/package.json]: We may want to set `"private": true` so we don't publish by default since we'll always want to ensure we have the correct version and tag, not `latest`. We can override in our release scripts with `publishConfig`.

**Comment by simi** [comment]
*Just being curious what's the usecase for this?*

**Comment by schneidmaster** [comment]
@simi Using ActionCable in standalone frontend applications that interact with a Rails API. (React, Angular, Ember, etc)

File: tasks/release.rb (hunk 1 above)
Comment (jeremy) [inline tasks/release.rb]: Take a look at https://rubygems.org/gems/rails/versions and https://github.com/rails/rails/releases to see our version style in history and at https://github.com/rails/rails/blob/v4.1.14.2/version.rb to see a 4-number version in practice.

**Comment by jeremy** [comment]
Also: using Action Cable in your Rails app with an alternative asset pipeline.

**Comment by schneidmaster** [comment]
PR'd into @maclover7's branch to add a Gulp build system (which also produces a usable compiled JS output file 😄). https://github.com/maclover7/rails/pull/1

**Comment by simi** [comment]
Thanks for info @schneidmaster and @jeremy.  Isn't https://github.com/rails/rails/blob/master/RELEASING_RAILS.md#release-the-gem change needed also to keep it updated?

**Comment by schneidmaster** [comment]
Not really -- the `npm publish` is baked into `tasks/release.rb` in this PR.

**Comment by simi** [comment]
Yup, but shouldn't be `npm publish` part of push task?

File: tasks/release.rb [hunk 13]
```diff
@@ -44,13 +44,21 @@
...
+        end
+      end
     end
 
     task gem => %w(update_versions pkg) do
       cmd = ""
       cmd << "cd #{framework} && " unless framework == "rails"
       cmd << "bundle exec rake package && " unless framework == "rails"
       cmd << "gem build #{gemspec} && mv #{framework}-#{version}.gem #{root}/pkg/"
+      cmd << "&& npm publish" if File.exist?("#{framework}/package.json")
```
Comment (jeremy) [inline tasks/release.rb]: Building the gem should not publish it to NPM. No more than it should push the gem to RubyGems. This is a _build_ task, not a _publish_ task.

**Comment by jeremy** [comment]
@simi Absolutely right. Building and publishing are separate tasks. We need to set the npm publish `--tag` too.

**Comment by simi** [comment]
@schneidmaster I think [it (RELEASING_RAILS.md)](https://github.com/rails/rails/blob/master/RELEASING_RAILS.md#release-the-gem) needs change since there are mentioned commands used in release and publish tasks and `npm publish` is missing there.

File: actioncable/package.json (hunk 10 above)
Comment (maclover7) [inline actioncable/package.json]: Neat, didn't know they had shortcuts for this 👍

File: actioncable/package.json (hunk 11 above)
Comment (maclover7) [inline actioncable/package.json]: We're going to be adding a `test` script here, but since that's not being added here, I'll ✂️

File: actioncable/package.json (hunk 12 above)
Comment (maclover7) [inline actioncable/package.json]: How would publishing work then, if we set `"private": true`? `npm publish` would fail: `If you set "private": true in your package.json, then npm will refuse to publish it.`

**Comment by maclover7** [comment]
@jeremy why do we need to run `npm publish --tag {version here}`, when it will by default take in the version number from the package.json (via `npm publish`)? Seems like an extra step 😬  (thought I am a bit of an npm novice, so do correct me if I'm wrong 😄)

**Comment by rafaelfranca** [comment]
Please let's keep our release simple. No new commands to run neither gulp,
grunt, webpack, whatever javascript dependency.

**Comment by schneidmaster** [comment]
@rafaelfranca I refactored my work to avoid the Gulp dependency, pending an issue with Sprockets that I'm trying to work out.

**Comment by jeremy** [comment]
@maclover7 Just need to be aware of them. Default tag is `latest` for any npm publish/install, but you can use e.g. `beta`, `unstable`, `release`, whatever, too.

File: tasks/release.rb [hunk 14]
```diff
@@ -44,6 +44,13 @@
       raise "Could not insert PRE in #{file}" unless $1
 
       File.open(file, 'w') { |f| f.write ruby }
+
+      if File.exist?("#{framework}/package.json")
+        Dir.chdir("#{framework}") do
+          version = version.gsub(/\./).with_index { |s, i| i >= 2 ? '-' : s }
```
Comment (rafaelfranca) [inline tasks/release.rb]: what is `s` and what is `i`? Also what this line does?

Comment (rafaelfranca) [inline tasks/release.rb]: For what i could get this is going to generate the wrong version:

```irb
>> version = "5.0.1.1"
=> "5.0.1.1"
>> puts version.gsub(/\./).with_index { |s, i| i >= 2 ? '-' : s }
5.0.1-1
```

File: tasks/release.rb [hunk 15]
```diff
@@ -61,6 +68,7 @@
 
     task :push => :build do
       sh "gem push #{gem}"
+      sh "npm publish" if File.exist?("#{framework}/package.json")
```
Comment (rafaelfranca) [inline tasks/release.rb]: This mean that who is releasing need to have npm installed?

File: tasks/release.rb (hunk 14 above)
Comment (jeremy) [inline tasks/release.rb]: @rafaelfranca see https://github.com/rails/rails/pull/24903#discussion_r62425201

@maclover7 Again, a helper method would eliminate this question ^ for anyone reading the code.

Comment (jeremy) [inline tasks/release.rb]: (Yes, it is incomplete.)

File: tasks/release.rb (hunk 15 above)
Comment (jeremy) [inline tasks/release.rb]: We'll also need similar npm ownership sharing like we do for gem owners.

Comment (rafaelfranca) [inline tasks/release.rb]: Seems we are not owners https://www.npmjs.com/package/actioncable

Comment (jeremy) [inline tasks/release.rb]:
```bash
$ npm owner ls actioncable
maclover7 <me@jonathanmoss.me>
```

Comment (maclover7) [inline tasks/release.rb]: I pinged @jeremy on Basecamp to ask if he could help me with getting Core setup with npm accounts, and I never heard back 😬

File: tasks/release.rb (hunk 14 above)
Comment (maclover7) [inline tasks/release.rb]: Added documentation to this area of `tasks/release.rb`.

File: tasks/release.rb (hunk 15 above)
Comment (maclover7) [inline tasks/release.rb]: Put out a post on the message board, for people to register for npm and send me their usernames

File: actioncable/Rakefile [hunk 16]
```diff
@@ -34,7 +34,7 @@ end
 
 namespace :assets do
   root_path = Pathname.new(dir)
-  destination_path = root_path.join("lib/assets/compiled")
+  destination_paths = [ root_path.join("lib/assets/compiled"), root_path.join("dist") ]
```
Comment (javan) [inline actioncable/Rakefile]: Why two destination paths? Can we point package.json to `lib/assets/compiled/*` instead?

File: RELEASING_RAILS.md [hunk 17]
```diff
@@ -125,6 +125,16 @@ $ git push --tags
 $ for i in $(ls pkg); do gem push $i; done
 ```
 
+However, due to that Action Cable also maintains an `npm` package (you must have `npm` installed in order to release Rails) here are the additional build steps that are run to release it:
+
+```
+# This is run as apart of the actioncable:update_versions rake task
```
Comment (jeremy) [inline RELEASING_RAILS.md]: `apart` → `a part`

File: RELEASING_RAILS.md [hunk 18]
```diff
@@ -125,6 +125,16 @@ $ git push --tags
...
+npm version #{version} --no-git-tag-version
+
+# This is run as apart of the actioncable:push rake task
```
Comment (jeremy) [inline RELEASING_RAILS.md]: Ditto

**Comment by javan** [comment]
Since we only exports a single global (`window.ActionCable`), we're guaranteed to receive "issues" re: not working with WebPack, Browserify, etc. See https://github.com/basecamp/trix/issues/119, https://github.com/basecamp/trix/issues/80, https://github.com/turbolinks/turbolinks/issues/91, https://github.com/turbolinks/turbolinks/issues/13 😁. It's probably worth including WebPack instructions like https://github.com/turbolinks/turbolinks#installation-using-webpack, but that won't please everyone.

File: actioncable/Rakefile (hunk 16 above)
Comment (maclover7) [inline actioncable/Rakefile]: Removed `dist/` -- not needed.

File: actioncable/package.json [hunk 19]
```diff
@@ -0,0 +1,24 @@
+{
+  "name": "actioncable",
+  "version": "5.0.0-rc1",
+  "description": "WebSocket framework for Ruby on Rails.",
+  "main": "lib/assets/compiled/action_cable.js",
+  "files": [
+    "lib/assets/compiled//*.js"
```
Comment (javan) [inline actioncable/package.json]: One too many slashes there?

Comment (maclover7) [inline actioncable/package.json]: 👍 ✂️

File: actioncable/app/assets/javascripts/action_cable.coffee.erb [hunk 20]
```diff
@@ -33,3 +33,12 @@
...
       console.log("[ActionCable]", messages...)
+
+# Since we are `module.exports`-ing below, in order to make the internal `ActionCable class`
+# recognize the `ActionCable` constant, we need to place the constant on the `window`, so that
+# way it will be recognized.
+window.ActionCable = @ActionCable
+
+# We need to module.exports here, in order to make ActionCable a "true Node.js module".
+# This makes it so that way an `Import` or `require` will load in the `ActionCable` module.
+module?.exports = @ActionCable
```
Comment (jeremy) [inline actioncable/app/assets/javascripts/action_cable.coffee.erb]: Setting this up is a separate concern, too. Let's do a separate PR for module loading compatibility. Targeting node.js itself seems… unlikely, huh? We're targeting asset bundlers (webpack, browserify, etc) that use module loading, so we'll want to identify which loaders to support. CommonJS, AMD, browser globals, ?  Separate PR :)

Comment (schneidmaster) [inline actioncable/app/assets/javascripts/action_cable.coffee.erb]: Jon actually did this as a result of our ongoing conversation about how to expose the module to NPM -- he was just concatenating the JS files into dist but that didn't work because the `ActionCable` var was undefined and as a result trying to use the NPM package would just auto-error. It's worth having a discussion about which module loaders to support, but IMO this is essential to the PR even doing anything useful.

Comment (maclover7) [inline actioncable/app/assets/javascripts/action_cable.coffee.erb]: Talked this over with @javan -- in order for the npm package to actually be used, we need to conform to npm's way of module loading. Without this, the package would be totally useless (no way to actually get at the `ActionCable` constant 😞)

**Comment by jeremy** [comment]
Squashed 548c1d6e8b819ca4e02e6218b67107c580ee65f2

**Comment by maclover7** [comment]
:metal:
